gh-129288: Add optional l2_cid and l2_bdaddr_type in BTPROTO_L2CAP socket address tuple

themadinventor opened the pull request:
To be able to connect L2CAP sockets to Bluetooth LE devices, the `l2_bdaddr_type` field in `sockaddr_l2` must be set to `BDADDR_LE_PUBLIC` or `BDADDR_LE_RANDOM`.
Likewise, when opening a raw LE L2CAP socket to the ATT service, `l2_cid` must be set instead of `l2_psm`.

This change adds support for providing l2_cid and l2_bdaddr_type as optional, traliing elements in the address tuple passed to `connect()`


<!-- gh-issue-number: gh-129288 -->
* Issue: gh-129288
<!-- /gh-issue-number -->

themadinventor committed:
gh-129288: Add optional bdaddr_type in BTPROTO_L2CAP socket address tuple

To be able to connect L2CAP sockets to Bluetooth LE devices, the l2_bdaddr_type must be set to BDADDR_LE_PUBLIC or BDADDR_LE_RANDOM.
This change adds support for providing the l2_bdaddr_type as an optional, traliing element in the address tuple passed to connect()

themadinventor committed:
gh-129288: Expose socket.BDADDR_BREDR, socket.BDADDR_LE_PUBLIC, socket.BDADDR_RANDOM only if defined for target

Add check that BDADDR_BREDR is defined as Windows build may set USE_BLUETOOTH but not BDADDR_*

themadinventor committed:
gh-129288: Add optional CID field in BTPROTO_L2CAP socket address tuple

This change adds support for the CID field in the socket address tuple. This allows e.g raw LE ATT connections.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

themadinventor committed:
gh-129288: Fix NEWS blurb formatting

themadinventor committed:
Merge branch 'main' into gh-129288

rruuaanng commented on Modules/socketmodule.c:
```diff
@@ -1493,9 +1493,18 @@ makesockaddr(SOCKET_T sockfd, struct sockaddr *addr, size_t addrlen, int proto)
+                                        addrobj,
+                                        _BT_L2_MEMB(a, psm),
+                                        _BT_L2_MEMB(a, cid),
+                                        _BT_L2_MEMB(a, bdaddr_type));
```
A tip, this can reduce one nesting and an assignment statement. WDYT?
```suggestion
if (addrobj) {
      /* Retain old format for non-LE address */
      ret = _BT_L2_MEMB(a, bdaddr_type) == BDADDR_BREDR
            ? Py_BuildValue("Oi", addrobj, _BT_L2_MEMB(a, psm))
            : Py_BuildValue("OiiB", addrobj,
				    _BT_L2_MEMB(a, psm),
				    _BT_L2_MEMB(a, cid),
				    _BT_L2_MEMB(a, bdaddr_type));
}
```

themadinventor replied:
Thank you for the suggestion! It is indeed more compact, however I personally also find it less readable. I'll let the CPython reviewers decide if your suggestion is more inline with their desired code style.

encukou replied:
IMO the expanded one is more readable, especially if we need to add new cases later.

As someone who's not familiar with these structs, I see a missing case when `bdaddr_type` is `BDADDR_BREDR` but `cid` is set. Could you add a comment if that's meaningless, or an assert if it's impossible?

themadinventor committed:
gh-129288: Update docs for socket to include new BTPROTO_L2CAP address fields

Update documentation to mention the new cid and bdaddr_type fields for BTPROTO_L2CAP connections.

themadinventor committed:
Merge branch 'main' into gh-129288

encukou commented on Modules/socketmodule.c:
```diff
@@ -7799,6 +7811,11 @@ socket_exec(PyObject *m)
+#ifdef BDADDR_BREDR
+    ADD_INT_MACRO(m, BDADDR_BREDR);
+    ADD_INT_MACRO(m, BDADDR_LE_PUBLIC);
+    ADD_INT_MACRO(m, BDADDR_LE_RANDOM);
```
Please add these to `Doc/library/socket.rst`, see existing `BDADDR_ANY` docs for an example.

encukou commented on Doc/library/socket.rst:
```diff
@@ -137,8 +137,9 @@ created.  Socket addresses are represented as follows:
-    the Bluetooth address as a string and ``psm`` is an integer.
+  - :const:`BTPROTO_L2CAP` accepts ``(bdaddr, psm, cid, bdaddr_type)`` where ``bdaddr`` is
+    the Bluetooth address as a string and ``psm``, ``cid`` and ``bdaddr_type`` are integers.
+    ``cid`` and ``bdaddr_type`` are optional.
```
Please add a ``.. versionchanged:: next`` note. (Look around for examples, but use the literal word `next` instead of "3.14".)

encukou commented on Modules/socketmodule.c:
```diff
@@ -7799,6 +7811,11 @@ socket_exec(PyObject *m)
 #endif /* BTPROTO_RFCOMM */
     ADD_STR_CONST(m, "BDADDR_ANY", "00:00:00:00:00:00");
     ADD_STR_CONST(m, "BDADDR_LOCAL", "00:00:00:FF:FF:FF");
+#ifdef BDADDR_BREDR
```
The code above doesn't use `#ifdef BDADDR_BREDR`. What's the right guard -- is `BTPROTO_L2CAP` enough (in which case it should be used here) or should `#ifdef BDADDR_BREDR` be added above?

themadinventor replied:
I've now moved those consts inside the `BTPROTO_L2CAP` guard in [2c537ba](https://github.com/python/cpython/pull/129293/commits/2c537ba7f5394357ae9f457af7b2787af0971155).

The `BDADDR_BREDR` and friends were added to Linux and Bluez some years after `BTPROTO_L2CAP`, so in theory using the `BDADDR_BREDR` guard would allow compiling on systems within that time window.

However, `BDADDR_BREDR` was added ~13 years ago [1], and I don't know if CPython is striving to support systems that old with new releases. If we'd like to, then we'd also have to add guards to the `sockaddr_l2` routines, as the `l2_cid` and `l2_bdaddr_type` struct members were added around the same time.

[1] https://github.com/bluez/bluez/commit/ba801fcd67e4fa5169e8c8b73fd87b6c08fca335

encukou replied:
No, 13-year-old systems aren't relevant for Python 3.14.
Should we be concerned about non-Linux systems, though? Are you aware of any BlueZ “lookalikes” that might define `BTPROTO_L2CAP` without `BDADDR_BREDR`?

themadinventor replied:
It looks like FreeBSD is safe, having introduced the struct fields and `BDADDR_BREDR`... macros back in 2015. NetBSD is looks problematic, with `BTPROTO_L2CAP` defined but no LE support, afaict.

However, it seems like Bluetooth socket support is disabled on NetBSD? There's a `USE_BLUETOOTH` guard, and it looks like it's not enabled on NetBSD or DragonFly (line 478). Perhaps they patch it downstream, or maybe it's just getting late, and I'm not parsing those lines correctly :sweat_smile: 

I don't run any of those systems unfortunately, so I can't just test it easily.

encukou replied:
OK! Let's assume all systems with `BTPROTO_L2CAP` have `BDADDR_BREDR`. Thanks!

encukou reviewed (commented):
Thank you!
I guess we don't have active Bluetooth experts, so I'll look at this mostly as a generic C dev.
Are you aware of some official docs that would explain what, for example, what a `cid` is? `man l2cap` doesn't help much.

Are there any tests you could add to `Lib/test/test_socket.py`?

themadinventor committed:
gh-129288: Improve docs for socket on BTPROTO_L2CAP, BDADDR_BREDR et al

themadinventor committed:
gh-129288: Move BDADDR_BREDR et al macros to use BTPROTO_L2CAP guard

encukou committed:
Expand the docs & provide link targets for the constants

themadinventor committed:
gh-129288: Update comment re discarding cid for BR/EDR L2CAP socket

themadinventor committed:
Merge pull request #1 from encukou/gh-129288

BTPROTO_L2CAP extenstion: Expand the docs & provide link targets for the constants

themadinventor committed:
gh-129288: Add tests for BTPROTO_L2CAP socket

themadinventor committed:
gh-129288: Add HAVE_SOCKET_BLUETOOTH_L2CAP to socket tests

themadinventor committed:
gh-129288 Revert skip criteria for testCreateL2capSocket

encukou commented on Doc/library/socket.rst:
```diff
@@ -137,8 +137,19 @@ created.  Socket addresses are represented as follows:
+    - ``psm`` Integer specifying the Protocol/Service Multiplexer.
+    - ``cid`` Optional integer specifying the Channel Identifier. If not given,
+      defaults to zero.
+    - ``bdaddr_type`` Optional integer specifying the address type;
```
```suggestion
    - ``bdaddr`` is a string specifying the Bluetooth address.
    - ``psm`` is an integer specifying the Protocol/Service Multiplexer.
    - ``cid`` is an optional integer specifying the Channel Identifier.
      If not given, defaults to zero.
    - ``bdaddr_type``is an optional integer specifying the address type;
```

encukou reviewed (commented):
Looks good!
One more wording touchup for the field docs, which follow the `where:`

encukou committed:
Update Doc/library/socket.rst

encukou commented on Doc/library/socket.rst:
```diff
@@ -137,8 +137,19 @@ created.  Socket addresses are represented as follows:
+    - ``psm`` is an integer specifying the Protocol/Service Multiplexer.
+    - ``cid`` is an optional integer specifying the Channel Identifier.
+      If not given, defaults to zero.
+    - ``bdaddr_type``is an optional integer specifying the address type;
```
```suggestion
    - ``bdaddr_type`` is an optional integer specifying the address type;
```

encukou committed:
Never commit from GitHub UI, right?
